### PR TITLE
[core][android]Decouple offline storage vacuum and the deleteRegion API

### PR DESCRIPTION
**Pull request by pozdnyakov**
Fixes https://github.com/mapbox/mapbox-gl-native-team/issues/102

### CODE REVIEW

File: platform/default/src/mbgl/storage/offline_database.cpp [hunk 1]
```diff
@@ -1298,6 +1299,15 @@ void OfflineDatabase::markUsedResources(int64_t regionID, const std::list<Resour
     handleError(ex, "mark resources as used");
 }
 
+std::exception_ptr OfflineDatabase::pack() try {
+    if (!db) initialize();
+    vacuum();
+    return nullptr;
+} catch (const mapbox::sqlite::Exception& ex) {
```
Comment (alexshalamov) [inline platform/default/src/mbgl/storage/offline_database.cpp]: catch generic exception? Initialize can throw IOException

Comment (pozdnyakov) [reply platform/default/src/mbgl/storage/offline_database.cpp]: nice catch! seems it's missing across the storage, I'll make a separate pr to handle it.

**Comment by pozdnyakov** [comment]
@LukasPaczos @tobrun could you pls :eyes: at the android-related changes, especially [OfflineRegion.deleteAndSkipPackDatabase](https://github.com/mapbox/mapbox-gl-native/pull/15899/commits/a60333a5fe172d309a13b4e7efd213928973bef7) API. Is this API shape quite appropriate for Android, would it be better to just add a flag to the existing `OfflineRegion.delete()` method (and thus save some binary size)?

File: include/mbgl/storage/default_file_source.hpp [hunk 2]
```diff
@@ -131,6 +131,16 @@ class DefaultFileSource : public FileSource {
      */
     void deleteOfflineRegion(OfflineRegion&&, std::function<void (std::exception_ptr)>);
 
+    /*
+     * Same as the method above but skipping database file packing for performance reasons.
+     *
+     * After this operation is complete, the database file does not shrink.
+     * Database file packing can be done later with `packDatabase()`.
+     *
+     * This method is useful optimization e.g. when several regions should be deleted in a row.
```
Comment (tmpsantos) [inline include/mbgl/storage/default_file_source.hpp]: Nit: `is a useful`

File: platform/default/src/mbgl/storage/default_file_source.cpp [hunk 3]
```diff
@@ -82,9 +82,9 @@ class DefaultFileSource::Impl {
         }
     }
 
-    void deleteRegion(OfflineRegion&& region, std::function<void (std::exception_ptr)> callback) {
+    void deleteRegion(OfflineRegion&& region, bool skipPack, std::function<void(std::exception_ptr)> callback) {
```
Comment (tmpsantos) [inline platform/default/src/mbgl/storage/default_file_source.cpp]: I just don't like the fact this is a double negative.

`false` here means doing pack. What about just `doPack` or `pack`?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java [hunk 4]
```diff
@@ -429,6 +429,53 @@ public void run() {
     }
   }
 
+  /**
+   * Same as the method above but skipping database file packing for performance reasons.
```
Comment (LukasPaczos) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java]: Could you link the method instead with `{@link com.mapbox.mapboxsdk.offline.OfflineRegion#delete(OfflineRegion.OfflineRegionDeleteCallback)}`?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java [hunk 5]
```diff
@@ -429,6 +429,53 @@ public void run() {
...
+   * <p>
+   * This method is useful optimization e.g. when several regions should be deleted in a row.
```
Comment (LukasPaczos) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java]: Would be great to highlight here that the database file size will not shrink and you need to call `packDatabase` to do it later (and link the method like above).

**Comment by pozdnyakov** [comment]
@LukasPaczos @tmpsantos Thanks for your comments! Fixed now in the latest patch set, PTAL.

File: include/mbgl/storage/default_file_source.hpp [hunk 6]
```diff
@@ -129,7 +134,17 @@ class DefaultFileSource : public FileSource {
...
+
+    /*
+     * Same as the method above but skipping database file packing for performance reasons.
+     *
+     * After this operation is complete, the database file does not shrink.
+     * Database file packing can be done later with `packDatabase()`.
+     *
+     * This method is a useful optimization e.g. when several regions should be deleted in a row.
+     */
+    void deleteOfflineRegionSkipPackDatabase(OfflineRegion&&, std::function<void(std::exception_ptr)>);
```
Comment (tmpsantos) [inline include/mbgl/storage/default_file_source.hpp]: leftover

Comment (pozdnyakov) [reply include/mbgl/storage/default_file_source.hpp]: Thanks for the catch!

**Pull request merged**